packages/block-directory/src/store/test/actions.js

```diff
@@ -25,288 +64,218 @@ describe( 'actions', () => {
 			],
 		},
 	};
-	const plugin = {
+
+	const pluginResponse = {
 		plugin: 'block/block.php',
 		status: 'active',
 		name: 'Test Block',
 		version: '1.0.0',
 		_links: {
-			self: [
-				{
-					href: pluginEndpoint,
-				},
-			],
+			self: [ { href: pluginEndpoint } ],
 		},
 	};
 
 	describe( 'installBlockType', () => {
-		const block = item;
-		it( 'should install a block successfully', () => {
-			const generator = installBlockType( block );
-
-			expect( generator.next().value ).toEqual( {
-				type: 'CLEAR_ERROR_NOTICE',
-				blockId: block.id,
-			} );
-
-			expect( generator.next().value ).toEqual( {
-				type: 'SET_INSTALLING_BLOCK',
-				blockId: block.id,
-				isInstalling: true,
-			} );
-
-			expect( generator.next().value ).toMatchObject( {
-				type: 'API_FETCH',
-				request: {
-					path: 'wp/v2/plugins',
-					method: 'POST',
-				},
-			} );
-
-			expect( generator.next( plugin ).value ).toEqual( {
-				type: 'ADD_INSTALLED_BLOCK_TYPE',
-				item: {
-					...block,
-					links: {
-						...block.links,
-						self: [
-							{
-								href: pluginEndpoint,
-							},
-						],
-					},
-				},
-			} );
-
-			expect( generator.next().value ).toEqual( {
-				type: 'LOAD_ASSETS',
-				assets: block.assets,
-			} );
-
-			expect( generator.next().value ).toEqual( {
-				args: [],
-				selectorName: 'getBlockTypes',
-				storeKey: blocksStore.name,
-				type: '@@data/SELECT',
-			} );
-
-			expect( generator.next( [ block ] ).value ).toMatchObject( {
-				type: '@@data/DISPATCH',
-				actionName: 'createInfoNotice',
-				storeKey: noticesStore.name,
-			} );
-
-			expect( generator.next().value ).toEqual( {
-				type: 'SET_INSTALLING_BLOCK',
-				blockId: block.id,
-				isInstalling: false,
-			} );
-
-			expect( generator.next() ).toEqual( {
-				value: true,
-				done: true,
-			} );
+		it( 'should install a block successfully', async () => {
+			const registry = createRegistryWithStores();
+
+			// mock the api-fetch and load-assets modules
+			apiFetch.mockImplementation( async ( { path } ) => {
+				switch ( path ) {
+					case 'wp/v2/plugins':
+						return pluginResponse;
+					default:
+						throw new Error( `unexpected API endpoint: ${ path }` );
+				}
+			} );
+
+			loadAssets.mockImplementation( loadAssetsMock( registry ) );
+
+			// install the block
+			await registry
+				.dispatch( blockDirectoryStore )
+				.installBlockType( block );
+
+			// check that blocks store contains the new block
+			const registeredBlock = registry
+				.select( blocksStore )
+				.getBlockType( 'block/block' );
+			expect( registeredBlock ).toBeTruthy();
```

Comment:
I understand why you did that, and at the same time testing an action by calling selectors feels off to me. A failed test could indicate a problem with any function called here. What's your take on mocking dispatch and registry calls?

Reply:
I have the same question for other tests; let's keep the discussion contained here.

Reply:
> What's your take on mocking dispatch and registry calls?

I think that not a good way to unit test things and I strongly prefer to test the entire store as one unit. The actions, reducers, resolvers and selectors are not very interesting as individual units -- it's their combination into a store where the interesting and test-worthy behavior emerges.

When mocking `dispatch` and everything else, I'm basically testing only that the action function calls other functions, but I already know that from reading the source code. I'd rather test what the function _returns_ and/or what _effects_ does the call have.

Reply:
I see, what we're really discussing is what _unit_ is being tested here. If I think about the store, not a specific action, it makes much more sense. And we have unit tests for these selectors, so should be easy to identify where did the failure originate.

Reply:
Testing a selector as a standalone unit makes sense, because it converts a state object to a return value and I just need to supply the state object. Similarly, reducer takes a state and an action and produces a new state. That's also self-contained even without a store.

But a complex action thunk that performs multiple operations on several registered stores? To really test what's the observable _result_ of that action, I need a realistic registry with the stores.

To show an analogy, consider a function that converts an array into another array. I can test it by passing some testing array as an argument and checking if the returned array is as expected. That's the common approach. Or I could test it by mocking the `Array` class and its methods, and then checking that the tested function called `map`, `push`, `slice` and `sort` in the expected order and with expected arguments. That would surprise most people and yet that's exactly how we're used to test `@wordpress/data` actions.

Or a function that uses the DOM API (`createElement`, `insertBefore`, `append`, ...) to add some HTML markup to the `document`. One way to test is to use `jsdom` and then check if the created markup has the structure we want (like `<h1><b>hello</b> world</h1>`). Another way is to mock the DOM APIs and check if the mocks were called in the expected order and the expected number of times. Again, the first approach is usually the preferred one.

Reply:
You are full of great insights @jsnajdr, thank you for elaborating! I'll borrow some of that code to test the other stores